lib/wallaroo/core/source/connector_source/connector_framed_source_notify.pony

```diff
@@ -82,91 +254,348 @@ class ConnectorSourceNotify[In: Any val]
+        ingest_ts, pipeline_time_spent)
+    end
+
+  /////////////////////////////
+  // Connector Message Handling
+  /////////////////////////////
+
+  fun ref received_connector_msg(source: ConnectorSource[In] ref,
+    data: Array[U8] iso,
+    latest_metrics_id: U16,
+    ingest_ts: U64,
+    pipeline_time_spent: U64): Bool
+  =>
+    if _prep_for_rollback then
+      // Anything that the connector sends us is ignored while we wait
+      // for the rollback to finish.  Tell the connector to restart later.
+      send_restart()
+      return _continue_perhaps(source)
+    end
+
+    _credits = _credits - 1
+    if (_credits <= _refill_credits) and
+        (_fsm_state is _ProtoFsmStreaming) then
+      // Our client's credits are running low and we haven't replenished
+      // them after barrier_complete() processing.  Replenish now.
+      _send_acks()
+    end
+
+    try
+      let data': Array[U8] val = consume data
       ifdef "trace" then
-        @printf[I32](("Rcvd msg at " + _pipeline_name + " source\n").cstring())
+        @printf[I32]("NH: decode data: %s\n".cstring(), _print_array[U8](data').cstring())
       end
+      let connector_msg = cwm.Frame.decode(consume data')?
+      match connector_msg
+      | let m: cwm.HelloMsg =>
+        ifdef "trace" then
+          @printf[I32]("TRACE: got HelloMsg\n".cstring())
+        end
+        if _fsm_state isnt _ProtoFsmConnected then
+          @printf[I32]("ERROR: %s.received_connector_msg: state is %d\n".cstring(),
+            __loc.type_name().cstring(), _fsm_state())
+          Fail()
+        end
 
-      (let is_finished, let last_ts) =
-        try
-          let decoded =
-            try
-              _handler.decode(consume data)?
-            else
-              ifdef debug then
-                @printf[I32]("Error decoding message at source\n".cstring())
-              end
-              error
-            end
-          let decode_end_ts = WallClock.nanoseconds()
-          _metrics_reporter.step_metric(_pipeline_name,
-            "Decode Time in Connector Source", latest_metrics_id, ingest_ts,
-            decode_end_ts)
-          latest_metrics_id = latest_metrics_id + 1
+        if m.version != "0.0.1" then
+          @printf[I32]("ERROR: %s.received_connector_msg: unknown protocol version %s\n".cstring(),
+            __loc.type_name().cstring(), m.version.cstring())
+          return _to_error_state(source, "Unknown protocol version")
+        end
+        if m.cookie != _cookie then
+          @printf[I32]("ERROR: %s.received_connector_msg: bad cookie %s\n".cstring(),
+            __loc.type_name().cstring(), m.cookie.cstring())
+          return _to_error_state(source, "Bad cookie")
+        end
 
-          ifdef "trace" then
-            @printf[I32](("Msg decoded at " + _pipeline_name +
-              " source\n").cstring())
-          end
+        // SLF TODO: add routing logic to handle
+        // m.program_name and m.instance_name routing requirements
```

Comment:
What does `m.program_name` refer to?

Reply:
`m.program_name` and `m.instance_name` are parts of the v2 connector protocol description that aren't implemented yet (as far as I know) by a connector source or sink.